File: 730/D2_RPD_and_Rap_Sheet_Hard_Version_.cpp
#include<bits/stdc++.h>

using namespace std;

//typedef long long int;
#define inf 1e18
#define mod 1000000007
#define pb push_back

int xorr(int a,int b,int k){
    int p=1;
    int ans=0;
    while(a>0 or b>0){
        ans+=((a%k - b%k + k)%k)*p;
        a/=k;
        b/=k;
        p*=k;
    }
    return ans;
}

int main(int argc, char const *argv[])
{
    int T;
    cin>>T;
    
    for(int t=0; t<T; t++){
        int n;
        cin>>n;
        
        int k;
        cin>>k;
        
        
        int done =0;
        int i=0;
        while(done==0){
            if(i==0)
                cout<<0<<endl;
            else{
                if(i%2==0)
                    cout<<xorr(i,i-1,k)<<endl;
                else
                    cout<<xorr(i-1,i,k)<<endl;
            }
            cin>>done ;
            cout.flush();
            i+=1;
        }
    }
    return 0;
}

File: 730/D2_RPD_and_Rap_Sheet_Hard_Version_.py
import sys

def xor(a,b,k):
    p=1
    ans=0
    while(a>0 or b>0):
        ans+=((a%k - b%k + k)%k)*p
        a//=k
        b//=k
        p*=k
    return ans

for _ in range(int(input())):
    n,k =map(int,input().split())
    done =0
    i=0
    while(done==0):
        if(i==0):
            print(0)
        else:
            if(i%2==0):
                print(xor(i,i-1,k))
            else:
                print(xor(i-1,i,k))
        done = int(input())
        sys.stdout.flush()
        i+=1
                    

    